core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -105,16 +105,31 @@ class ReplicaFetcherThread(name:String,
       leaderEndOffset
     } else {
       /**
-       * The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
+       * If the leader's log end offset is greater than follower's log end offset, there are two possibilities:
+       * 1. The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
        * start offset because the leader has deleted old logs (log.logEndOffset < leaderStartOffset).
+       * 2. When leader migration occurs, there is a slight chance that the old leader's high watermark is greater than
+       * the new leader's log end offset, so an OffsetOutOfRangeException was thrown. After that some more messages are
+       * produced to the leader. When the follower was trying to handle the exception and query the log end offset of
+       * broker, the leader's log end offset become higher than follower's log end offset.
+       *
+       * In the first case, the follower's current log end offset is smaller than leader's log start offset. So the
+       * follower should truncate all its log, roll out a new segment and start to fetch from current leader's log
+       * start offset.
+       * In the second case, the follower should just keep the current log segments and retry fetch.
```

Comment:
@gwenshap Consider the following case:
1. Leader broker 0 has LEO = 100, follower broker 1 has LEO=90. The starting offset is 0. Follower broker 0 is not in ISR.
2. Leader migration occurs. Now broker 1 becomes leader and broker 0 becomes follower. The LEO on the new leader(broker 1) is 90, the high watermark on broker 0 is 100. (Unclean leader election) 
3. Broker 0 truncates its log to its high watermark 100 and started to fetch from broker 1 at offset 100.
4. Broker 1 returns an OffsetOutOfRangeExecption because 100 is greater than its LEO 90.
5. Broker 1 as the new leader appended some more message to the log and now its LEO become 105.
6. Broker 0 tries to handle the OffsetOutOfRagneException, so it queries the LEO of broker 1 again to find why the exception was thrown. Broker 0 found Broker 1's LEO is 105. Supposedly, in this case broker 0 should just fetch from offset 100 again. The messages at offsets 90-100 will be different on broker 0 and broker 1 but we live with it by design. Because the bug in the current code, broker 0 will think it received the OffsetOutOfRangeException because 100 was lower than the starting offset on broker 1. So instead of fetching from 100, broker 0 will truncate all its log to 0 and restart fetching all the logs.

So that is why in the second case we should just retry fetch because the previous OffsetOutOfRangeException is no longer valid.